Upload versioned connector docs on publish

lmossman opened the pull request:
## What
Resolves https://github.com/airbytehq/airbyte/issues/29400

This PR adds logic to the metadata_service which uploads connector documentation to GCS along with the other metadata uploads.

This will be utilized in an upcoming airbyte server API endpoint to retrieve docs for specific connector versions, which will be utilized in the UI to show the correct documentation for the connector version actively being used.

## How
Adds a `_docs_upload()` function to the `gcs_upload.py` file which does the following:
- Gets the path to the connector's doc file by walking up the metadata file path and then down to the docs folder, and using the connector name from the metadata's documentationUrl path to determine the name of the docs file
- Checks for both `.inapp.md` and `.md` files, and uploads both if they exist
- This function is called once with `latest=False` to upload to the versioned path, and once with `latest=True` to upload to the latest path, ensuring that doc files are present in both.
  - Having the doc file in the latest path is important as this will be used by the upcoming server API as a fallback if it cannot find a doc in the versioned path
  - As a follow-up to this PR, I will write and execute a script that uploads the current docs of every connector to the `latest` path so that the fallback is always present

**Note**: I decided to keep the connectors docs where they are today, as opposed to moving them next to the `metadata.yaml` files, because Sherif and I decided that we wanted to keep this change as small in scope as possible so that we can complete the other UI doc tasks we have planned for this quarter.

This small change will unlock the ability to have versioned docs in the UI, and once Connector Ops feels the need to prioritize it, they can do the full migration of these docs into the connector folders, updating docusaurus, etc. 

## 🚨 User Impact 🚨
No breaking changes

lmossman commented:
@bnchrch what is the best way to test this change on CI? Is there a slash command I can run try running the publishing step? I assume we have something like this for pre-releases

lmossman commented on airbyte-ci/connectors/metadata_service/lib/tests/test_commands.py:
```diff
@@ -51,29 +51,38 @@ def mock_metadata_upload_info(
-        (True, False, False),
-        (False, True, False),
-        (False, False, True),
-        (True, True, False),
-        (True, False, True),
-        (False, True, True),
-        (True, True, True),
+        (False, False, False, False, False, False, False),
+        (True, False, False, False, False, False, False),
+        (False, True, False, False, False, False, False),
+        (False, False, True, False, False, False, False),
+        (True, True, False, False, False, False, False),
+        (True, False, True, False, False, False, False),
+        (False, True, True, False, False, False, False),
+        (True, True, True, False, False, False, False),
+        (True, True, True, True, True, True, True),
     ],
 )
```
Including every permutation of 7 different boolean parameters would be 2^7 = 128 different test cases which felt unnecessary, so I just included one test case where all are set to true

bnchrch replied:
That seems fair!

erohmensing replied:
This might be helpful in this case (I've often found myself wanting the feature in java): 
<img width="932" alt="image" src="https://github.com/airbytehq/airbyte/assets/22333512/708ca9af-d81d-467f-98e2-bc44ae22a468">
 (From https://docs.pytest.org/en/latest/how-to/parametrize.html#pytest-mark-parametrize-parametrizing-test-functions)

erohmensing replied:
If indeed we are just enumerating all the options

bnchrch commented:
Hey @lmossman let me answer some of these larger questions!

Then ill be back for an indepth code review later.

> what is the best way to test this change on CI?

So I believe you can test using `airbyte-ci` safely so long as you have
1. `METADATA_SERVICE_BUCKET_NAME` pointed at your dev bucket
2. `METADATA_SERVICE_GCS_CREDENTIALS` pointed at a service account with access to that bucket
3. Are using a connector who docker image is already built and published to dockerhub (this should be all of them)

Heres an example of the env vars I set
```sh
export METADATA_SERVICE_GCS_CREDENTIALS=`cat ~/Development/secrets/dev-metadata-service-account.json`
export METADATA_SERVICE_BUCKET_NAME="dev-airbyte-cloud-connector-metadata-service"

export SPEC_CACHE_BUCKET_NAME="io-airbyte-cloud-spec-cache"
export SPEC_CACHE_GCS_CREDENTIALS=`cat ~/Development/secrets/prod-metadata-service-account.json`

export GCP_GSM_CREDENTIALS=`cat ~/Development/secrets/ben-dev-account-dataline-service-key.json`

export DOCKER_HUB_USERNAME="benairbyte"
export DOCKER_HUB_PASSWORD="REDACTED"
```

Afterwards you should be free to run
```sh
airbyte-ci connectors --name=source-faker test # which triggers metadata validate
airbyte-ci connectors --name=source-faker publish --main-release # which triggers metadata upload
```

> On markdown images
This is tricky!

So the main question lies with "who is responsible for turning relative paths to absolute urls?"

The FE already does this so it would not be a stretch to do it there again.

However I think a good set principals to follow is
1. The platform BE should be unaware of implementation details of the connectors and related systems
2. FE should be unaware of the implementation details of any external sysem (platform, connectors or otherwise)

With that in mind I (naively) think that
1. the metadata service should be responsible for finding and uploading relative images to the bucket
2. the registry orchestrator should be responsible for transforming them into absolute urls

Where this breaks down / requires extra thought is
* How often are images not colocated? i.e. are we seeing image paths that look like this `../../../../someotherpage/someothersection/tutorial_123.png`?

lmossman commented:
> How often are images not colocated? i.e. are we seeing image paths that look like this ../../../../someotherpage/someothersection/tutorial_123.png?

@bnchrch I just checked and there are actually only 8 connector docs containing images, and all of them are either full github URLs, or they point to the `../../.gitbook/assets/` folder, and they all seem to be unique to each connector:
<img width="647" alt="Screenshot 2023-09-13 at 4 59 51 PM" src="https://github.com/airbytehq/airbyte/assets/22731524/18bb3084-cffe-431d-a969-57e994072762">

So that should make the image uploading logic fairly straightforward, i.e.:
- Inspect the markdown file
- Find all images with relative paths
- Upload those images to the GCS bucket

bnchrch commented on airbyte-ci/connectors/metadata_service/lib/metadata_service/commands.py:
```diff
@@ -24,7 +24,28 @@ def log_metadata_upload_info(metadata_upload_info: MetadataUploadInfo):
+            color="green"
+        )
+    if metadata_upload_info.doc_inapp_latest_uploaded:
+        click.secho(
```
super nit: We're getting a little bit long in the ifs here

I know I started it 😅

But if we created a helper I would not be mad

```py
def log_if_uploaded(metadata_upload_info: MetadataUploadInfo, upload_success_key: str, upload_id_key: str, name: str):
    success = metadata_upload_info[upload_success_key]
    if success:
        path = metadata_upload_info[upload_id_key]
        click.secho(
            f"The {name} file for {metadata_upload_info.metadata_file_path} was uploaded to {path}.",
            color="green"
        )

log_if_uploaded(metadata_upload_info, "icon_uploaded", "icon_blob_id", "icon")
```

bnchrch commented on airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:
```diff
@@ -119,6 +138,29 @@ def _icon_upload(metadata: ConnectorMetadataDefinitionV0, bucket: storage.bucket
+    local_doc_path = docs_folder_path / f"{doc_file_name}.md"
+    local_inapp_doc_path = docs_folder_path / f"{doc_file_name}.inapp.md"
+
+    remote_doc_path = get_doc_remote_file_path(metadata.data.dockerRepository, "latest" if latest else metadata.data.dockerImageTag, False)
```
👏  very nice

bnchrch commented on airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:
```diff
@@ -119,6 +138,29 @@ def _icon_upload(metadata: ConnectorMetadataDefinitionV0, bucket: storage.bucket
+    if local_doc_path.exists():
+        doc_uploaded, doc_blob_id = upload_file_if_changed(local_doc_path, bucket, remote_doc_path)
+    else:
+        doc_uploaded, doc_blob_id = False, f"No doc found at {local_doc_path}"
```
❓  Do we want to enforce that they have docs?

lmossman replied:
Yeah I think this should be enforced, I will add an exception here

bnchrch replied:
Great. Lets run test on a few GA connectors so that were sure we didnt cause a regression :)

bnchrch commented on airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:
```diff
@@ -119,6 +138,29 @@ def _icon_upload(metadata: ConnectorMetadataDefinitionV0, bucket: storage.bucket
+    docs_folder_path = metadata_file_path.parents[3] / DOCS_FOLDER_PATH / f"{connector_type}s"
+    
+    local_doc_path = docs_folder_path / f"{doc_file_name}.md"
+    local_inapp_doc_path = docs_folder_path / f"{doc_file_name}.inapp.md"
```
💅  it seems like inapp docs are distinct. 

We should either
1. Move the variables for inapp next to where they are used.
2. Split the inapp upload and local upload into a "private" functions e.g. `_inapp_doc_upload`

I believe I prefer 2

lmossman replied:
I went ahead and just modified this function to take in one more `inapp: bool` parameter, to make it work for either one

bnchrch commented on airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:
```diff
@@ -189,4 +234,12 @@ def upload_metadata_to_gcs(
         icon_uploaded=icon_uploaded,
         metadata_file_path=str(metadata_file_path),
+        doc_version_uploaded=doc_version_uploaded,
+        doc_version_blob_id=doc_version_blob_id,
```
I made a comment about "hey we have a lot of ifs in the log function"

I think thats a symptom of the MetadataUploadInfo being flat.

@lmossman Do you have the capacity to move this into a more generic `upload_files` field?

It should go a long way in removing boiler plate

```py
# Updated models

@dataclass(frozen=True)
class MetadataUploadInfo:
    metadata_uploaded: bool
    metadata_file_path: str
    uploaded_files: List[UploadedFile]

class UploadedFile:
    id: str
    uploaded: bool
    description: str
    blob_id: Optional[str]

# Updated upload helpers

def _version_upload(metadata: ConnectorMetadataDefinitionV0, bucket: storage.bucket.Bucket, metadata_file_path: Path) -> UploadedFile:
    version_path = get_metadata_remote_file_path(metadata.data.dockerRepository, metadata.data.dockerImageTag)
    uploaded, blob_id = upload_file_if_changed(metadata_file_path, bucket, version_path, disable_cache=True)
    return UploadedFile(
        id="version_metadata"
        uploaded=uploaded,
        description="versioned metadata file",
        blob_id=blob_id,
    )

# Updated main method

version_metadata_upload = _version_upload(metadata, bucket, metadata_file_path)

return MetadataUploadInfo(
    uploaded=version_metadata_upload.uploaded or latest_metadata_upload.uploaded,
    metadata_file_path=str(metadata_file_path),
    uploaded_files=[
        version_metadata_upload,
        latest_metadata_upload,
    ]
)

# updated logger

for file in metadata_upload_info.uploaded_files:
    if file.uploaded:
        click.secho(
            f"The {file.description} file for {metadata_upload_info.metadata_file_path} was uploaded to {file.plob_id}.",
            color="green"
        )
```

bnchrch reviewed (approved):
@lmossman This is great! 

The only comment I have that is a "blocker" is the question around should we fail if there are no docs?

bnchrch commented:
Also @lmossman I like your approach with the images

Question on that.

What path will you put them at in the docs folder?

I would prefer a relative path like `../../../../folder1/folder2/folder3/image.jpg

to become `assets/image.jpg` or `assets/folder3/image.jpg`

Thoughts?

lmossman commented:
@bnchrch I've addressed your comments and finally got the tests working after quite a bit of trial and error 😅 

The only thing I haven't addressed is the images - I talked with Tim about this and we decided that since the frontend will still need to assume the location of the connector doc file within the docs folder in order to properly resolve relative links to other docs pages, we are just going to go with the frontend-github-url replacement approach that I laid out in [this comment](https://github.com/airbytehq/airbyte/pull/30410#issuecomment-1718200896) in order to keep the scope of this change small, and get to supporting versioned docs more quickly while still having time to complete the other FE projects we have planned for this quarter.

Once connector ops can prioritize a more thorough migration of all doc files _along with their images_ into the connector folders, that uploading logic will hopefully be simpler, so we don't want to invest the effort to make that work for this interim state.

Hopefully you are okay with this more minimal approach for now!

bnchrch commented:
Totally ok! Done not perfect perfectly applies here! Rereviewing now :)

bnchrch commented on airbyte-ci/connectors/metadata_service/lib/metadata_service/commands.py:
```diff
@@ -12,20 +12,12 @@
-            f"The icon file {metadata_upload_info.metadata_file_path} was uploaded to {metadata_upload_info.icon_blob_id}.", color="green"
-        )
+    for file in metadata_upload_info.uploaded_files:
+        if file.uploaded:
```
<3

bnchrch commented on airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:
```diff
@@ -175,18 +207,63 @@ def upload_metadata_to_gcs(
+                description="versioned inapp doc",
+                blob_id=doc_inapp_version_blob_id,
+            ),
+            UploadedFile(
```
💅 Ideally we return these from their respective upload functions.

@lmossman Do you think thats a reasonable change we can make?

lmossman replied:
@bnchrch the reason I did it this way was because in some cases we wet the `uploaded` and `blob_id` to `False, None` without calling the upload function. So if `UploadedFile` is returned from the function, then we would need to duplicate the `id` and the `description` in `upload_metadata_to_gcs` as well for those cases where the upload function isn't called, so doing it this way felt DRYer to me.

What do you think?

bnchrch reviewed (approved):
Overall this looks just fine to me!

Theres one comment that would be great if we can get it

But it is not a blocker so approving now 👍 

Thanks for all the context sharing and being so open to feedback @lmossman 💎

lmossman commented:
@bnchrch thanks for the quick review!

I tried running `airbyte-ci connectors --name=source-faker test` with the env vars set as you recommended, but I got a bunch of failures that look like this in the acceptance tests step:
```
==================================== ERRORS ====================================
__________ ERROR at setup of TestSpec.test_config_match_spec[inputs0] __________

anyio_backend = 'asyncio', args = (), kwargs = {'anyio_backend': 'asyncio'}
backend_name = 'asyncio', backend_options = {}
runner = 

    def wrapper(*args, anyio_backend, **kwargs):  # type: ignore[no-untyped-def]
        backend_name, backend_options = extract_backend_and_options(anyio_backend)
        if has_backend_arg:
            kwargs["anyio_backend"] = anyio_backend
    
        with get_runner(backend_name, backend_options) as runner:
            if isasyncgenfunction(func):
>               yield from runner.run_asyncgen_fixture(func, kwargs)

/usr/local/lib/python3.10/site-packages/anyio/pytest_plugin.py:68: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/local/lib/python3.10/site-packages/anyio/_backends/_asyncio.py:2094: in run_asyncgen_fixture
    self._loop.run_until_complete(f)
/usr/local/lib/python3.10/asyncio/base_events.py:649: in run_until_complete
    return future.result()
/usr/local/lib/python3.10/site-packages/anyio/_backends/_asyncio.py:2074: in fixture_runner
    retval = await agen.asend(None)
/app/connector_acceptance_test/conftest.py:160: in dagger_client
    async with dagger.Connection(config=dagger.Config(log_output=sys.stderr)) as client:
/usr/local/lib/python3.10/site-packages/dagger/connection.py:45: in __aenter__
    conn = await stack.enter_async_context(Engine(self.cfg))
/usr/local/lib/python3.10/contextlib.py:619: in enter_async_context
    result = await _cm_type.__aenter__(cm)
/usr/local/lib/python3.10/site-packages/dagger/engine/conn.py:55: in __aenter__
    return await anyio.to_thread.run_sync(self.start)
/usr/local/lib/python3.10/site-packages/anyio/to_thread.py:33: in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
/usr/local/lib/python3.10/site-packages/anyio/_backends/_asyncio.py:877: in run_sync_in_worker_thread
    return await future
/usr/local/lib/python3.10/site-packages/anyio/_backends/_asyncio.py:807: in run
    result = context.run(func, *args)
/usr/local/lib/python3.10/site-packages/dagger/engine/conn.py:45: in start
    return self.from_env() or self.from_cli()
/usr/local/lib/python3.10/site-packages/dagger/engine/conn.py:42: in from_cli
    return stack.enter_context(cli_session)
/usr/local/lib/python3.10/contextlib.py:492: in enter_context
    result = _cm_type.__enter__(cm)
/usr/local/lib/python3.10/site-packages/dagger/engine/cli.py:43: in __enter__
    conn = self._get_conn(proc)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = 
proc = 

    def _get_conn(self, proc: subprocess.Popen) -> ConnectParams:
        # TODO: implement engine session timeout (self.cfg.engine_timeout?)
        assert proc.stdout
        conn = proc.stdout.readline()
    
        # Check if subprocess exited with an error
        if ret := proc.poll():
            out = conn + proc.stdout.read()
            err = proc.stderr.read() if proc.stderr and proc.stderr.readable() else None
    
            # Reuse error message from CalledProcessError
            exc = subprocess.CalledProcessError(ret, " ".join(proc.args))
    
            msg = str(exc)
            detail = err or out
            if detail and detail.strip():
                # `msg` ends in a period, just append
                msg = f"{msg} {detail.strip()}"
    
>           raise SessionError(msg)
E           dagger.exceptions.SessionError: Failed to start Dagger engine session: Command '/root/.cache/dagger/dagger-0.6.4 session --label dagger.io/sdk.name:python --label dagger.io/sdk.version:0.6.4 --label dagger.io/sdk.async:true' returned non-zero exit status 1.
```

Have you seen that before?

bnchrch commented:
@lmossman Hmm youre not targeting colima in your cli by any chance are you?

lmossman commented:
> @lmossman Hmm youre not targeting colima in your cli by any chance are you?

@bnchrch I don't think so (I'm not even really sure how to do that?)

lmossman commented:
- Should add something to metadata_validator.py to ensure that the doc file exists

lmossman commented:
@bnchrch I've made the changes that we discussed yesterday, in [these commits](https://github.com/airbytehq/airbyte/pull/30410/files/2930c80ad4586f616a6c47a14cb8a56f4245cca4..91a78008f0536443308029e1c704c9ff5183169b#diff-8d2e2dfdaa33aa34583293472ad061493cf88a0ac2955e8bfe0cdc7480de10d1):
- Added a `doc_path` argument to the `upload` and `validate` commands in the metadata service
- Passed the doc_path into `validate_and_load` and `upload_metadata_to_gcs` through the `ValidatorOptions` object
- Updated `_doc_upload` to use that doc_path instead of inferring it from the metadata
- Added a test `doc.md` file and a pointer to it in `__init__.py` so that I could use that path in the tests
- Updated the tests to use that fixture path

As discussed I will leave it to you to update the MetadataUpload class in airbyte-ci accordingly!

bnchrch commented on airbyte-ci/connectors/metadata_service/lib/metadata_service/commands.py:
```diff
@@ -26,28 +26,30 @@ def metadata_service():
+    metadata_file_path = metadata_file_path if not metadata_file_path.is_dir() else metadata_file_path / METADATA_FILE_NAME
 
-    click.echo(f"Validating {file_path}...")
+    click.echo(f"Validating {metadata_file_path}...")
```
😍 Like this name change

bnchrch commented on airbyte-ci/connectors/metadata_service/lib/metadata_service/commands.py:
```diff
@@ -26,28 +26,30 @@ def metadata_service():
-def validate(file_path: pathlib.Path):
-    file_path = file_path if not file_path.is_dir() else file_path / METADATA_FILE_NAME
+@click.argument("metadata_file_path", type=click.Path(exists=True, path_type=pathlib.Path))
+@click.argument("doc_path", type=click.Path(exists=True, path_type=pathlib.Path))
```
❗️ We should mark this required

bnchrch replied:
and we need to update the readme with an example https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/metadata_service/lib/README.md?plain=1#L35

lmossman replied:
I've marked the metadata-file-path, doc-path, and bucket-name arguments as required for the `validate` and `upload` commands, and I've updated the validate example in the readme

bnchrch reviewed (changes requested):
A couple small changes! Working on the CI side of it now

bnchrch reviewed (approved):
The upstream PR for updating `airbyte-ci` should be all set.

Approving now so the path to merge is all clear.

Thanks for the work on this @lmossman

lmossman commented:
For anyone tagged here - this was caused by CI applying auto-formatting changes to this PR which for some reason changed a ton of files in my PR. Not sure why this happened but please feel free to ignore this PR
